src/unix/linux.c

```diff
@@ -2281,6 +2281,124 @@ uint64_t uv_get_available_memory(void) {
+           cgroup_trimmed);
+
+  /* Read cpu.max */
+  if (uv__slurp(path, buf, sizeof(buf)) >= 0) {
+    char max_usage_str[16];
+    sscanf(buf, "%15s %lld", max_usage_str, &constraint.period_length);
+    if (strcmp(max_usage_str, "max") == 0)
+      constraint.quota_per_period = LLONG_MAX;
+    else
+      constraint.quota_per_period = strtoll(max_usage_str, NULL, 10);
+  }
+
+  /* Construct the path to the cpu.weight file */
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.weight", cgroup_size,
+           cgroup_trimmed);
+
+  /* Read cpu.weight */
+  if (uv__slurp(path, buf, sizeof(buf)) >= 0) {
+    sscanf(buf, "%u", &weight);
+    constraint.proportions = (double)weight / 100;
+  }
+
+  return constraint;
+}
+
+static char* uv__cgroup1_find_cpu_controller(const char* cgroup,
+                                             int* cgroup_size) {
+  char* cgroup_cpu;
+
+  /* Seek to the cpu controller line. */
+  cgroup_cpu = strstr(cgroup, ":cpu,");
+
+  if (cgroup_cpu != NULL) {
+    /* Determine the length of the mount path. */
+    cgroup_cpu = cgroup_cpu + sizeof(":cpu,") - 1;
+    *cgroup_size = (int)strcspn(cgroup_cpu, "\n");
+  }
+
+  return cgroup_cpu;
+}
+
+static uv__cpu_constraint uv__get_cgroupv1_constrained_cpu(const char* cgroup) {
+  uv__cpu_constraint constraint = {0, 0, 0};
+  char path[256];
+  char buf[BUF_SIZE];
+  unsigned int shares;
+  int cgroup_size;
+  char* cgroup_cpu;
+
+  cgroup_cpu = uv__cgroup1_find_cpu_controller(cgroup, &cgroup_size);
+
+  /* Construct the path to the cpu.max file */
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.cfs_quota_us",
+           cgroup_size, cgroup_cpu);
+
+  if (uv__slurp(path, buf, sizeof(buf)) >= 0) {
+    sscanf(buf, "%lld", &constraint.quota_per_period);
+  }
+
+  /* Construct the path to the cpu.cfs_period_us file */
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.cfs_period_us",
+           cgroup_size, cgroup_cpu);
+
+  /* Read cpu.cfs_period_us */
+  if (uv__slurp(path, buf, sizeof(buf)) >= 0) {
+    sscanf(buf, "%lld", &constraint.period_length);
+  }
+
+  /* Construct the path to the cpu.shares file */
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.shares", cgroup_size,
+           cgroup_cpu);
+
+  /* Read cpu.shares */
+  if (uv__slurp(path, buf, sizeof(buf)) >= 0) {
+    sscanf(buf, "%u", &shares);
+    constraint.proportions = (double)shares / 1024;
+  }
+
+  return constraint;
+}
+
+uv__cpu_constraint uv__get_constrained_cpu(void) {
+  char cgroup[1024];
+  char path[256];
+  char controllers[256];
+
+  /* Read the cgroup from /proc/self/cgroup */
+  snprintf(path, sizeof(path), "/proc/self/cgroup");
+  if (uv__slurp(path, cgroup, sizeof(cgroup)) < 0) {
+    uv__cpu_constraint constraint = {0, 0, 0};
+    return constraint;
+  }
+
+  /* Check if the system is using cgroup v2 */
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/cgroup.controllers");
+  if (uv__slurp(path, controllers, sizeof(controllers)) >= 0)
```

Comment:
If you're not going to use the contents, just `access("/sys/fs/cgroup/cgroup.controllers", F_OK)`.

That said... what's the reason for using cgroup.controllers here? Everywhere else we look at the format of /proc/self/cgroup to figure out if we're dealing with cgroups v1 or v2.